source/plugin/event/connection.rst

```diff
@@ -0,0 +1,167 @@
+    org.spongepowered.api.event.network.ServerSideConnectionEvent.Disconnect
+    org.spongepowered.api.event.network.ServerSideConnectionEvent.Intent
+
+.. note::
+    The :javadoc:`ServerSideConnectionEvent` events were reworked in API 11 to address common usability pitfalls
+    and provide extra guarantees regarding threading and event ordering.
+
+Connection events are crucial part of plugins and their incorrect usage might be lead to
+data corruption, race conditions or memory leaks. This documentation goes over the common
+best practices that should be followed to ensure good experience.
+
+Async, Threading And Event Order
+================================
+Incoming connections can be fragile and their connection state can suddenly change. Connections are also
+handled asynchronously up to the point the connections authenticity has been verified. This allows plugins to
+load related user data from a database during :javadoc:`ServerSideConnectionEvent.Auth` event. Correctly handling
+these edge cases can prove to be difficult. To help with this, Sponge provides extra guarantees.
+
+After a plugin has observed connection event, Sponge ensures that events are ordered correctly and no more than
+single event is in-flight for that particular :javadoc:`ServerSideConnection`. You are also guaranteed to
+receive :javadoc:`ServerSideConnectionEvent.Disconnect` for all incoming connections that has got past
+:javadoc:`ServerSideConnectionEvent.Intent`.
+
+.. warning::
+    The ordering guarantees covered in this section does not apply to API versions prior to 11. You are also not
+    guaranteed to receive ``ServerSideConnectionEvent.Disconnect`` before the player has successfully joined
+    the world. Caches and proper housekeeping is recommended to handle any potential edge cases.
+
+.. note::
+    Sponge does not provide guarantees that only single authentication request is processed for one
+    particular user. This may happen due to a race conditions and slow connectivity. The plugin developer
+    must be aware of this caveat and potentially provide safety-checks and syncronization primitives.
+
+Example A
+~~~~~~~~~
+- Server processes incoming connection.
+- Event ``ServerSideConnectionEvent.Intent`` is fired.
+- Connection drops.
+- Event ``ServerSideConnectionEvent.Disconnect`` is fired.
+
+Example B
+~~~~~~~~~
+- Server processes incoming connection.
+- Event ``ServerSideConnectionEvent.Intent`` is fired.
+- Event ``ServerSideConnectionEvent.Auth`` is fired.
+    - Plugin A starts to load user data.
+    - Connection drops. Auth event is still being processed, disconnect event is queued.
+    - Plugin A finishes loading user data.
+- Event ``ServerSideConnectionEvent.Disconnect`` is fired.
+
+Best Practices
+==============
+Due to the nature of connection events being asynchronously, plugin developers **must** use thread-safe
+collections when storing, accessing and mutating related data.
```

Comment:
Due to many people not knowing there is thread safe collections, could we have it link to here (the official docs on concurrent collections). I know you have a example below, just a link to give more examples)

https://docs.oracle.com/javase/tutorial/essential/concurrency/collections.html